subprojects/core/src/main/java/org/gradle/api/internal/artifacts/JavaEcosystemAttributesDescriber.java

```diff
@@ -0,0 +1,299 @@
+            case LibraryElements.JAR:
+                sb.append("packaged as a jar");
+                break;
+            case LibraryElements.CLASSES:
+                sb.append("preferably in the form of class files");
+                break;
+            case LibraryElements.RESOURCES:
+                sb.append("preferably only the resources files");
+                break;
+            case LibraryElements.CLASSES_AND_RESOURCES:
+                sb.append("preferably not packaged as a jar");
+                break;
+            default:
+                sb.append("with the library elements '").append(name);
+        }
+    }
+
+    @SuppressWarnings("deprecation")
+    private static void describeUsage(Object usage, StringBuilder sb) {
+        String str = toName(usage);
+        switch (str) {
+            case Usage.JAVA_API:
+            case Usage.JAVA_API_CLASSES:
+            case Usage.JAVA_API_JARS:
+                sb.append("an API");
+                break;
+            case Usage.JAVA_RUNTIME:
+            case Usage.JAVA_RUNTIME_CLASSES:
+            case Usage.JAVA_RUNTIME_JARS:
+                sb.append("a runtime");
+                break;
+            default:
+                sb.append("a usage of '").append(str).append("'");
+        }
+    }
+
+    private static void describeTargetJvm(Object targetJvm, StringBuilder sb) {
+        sb.append("Java ").append(targetJvm);
+    }
+
+    private static void describeCategory(Object category, StringBuilder sb) {
+        String name = toName(category);
+        switch (name) {
+            case Category.LIBRARY:
+                sb.append("a library");
+                break;
+            case Category.REGULAR_PLATFORM:
+                sb.append("a platform");
+                break;
+            case Category.ENFORCED_PLATFORM:
+                sb.append("an enforced platform");
+                break;
+            case Category.DOCUMENTATION:
+                sb.append("documentation");
+                break;
+            default:
+                sb.append("a component of category '").append(name).append("'");
+        }
+    }
+
+    private static void describeDocsType(Object docsType, StringBuilder sb) {
+        String name = toName(docsType);
+        switch (name) {
+            case DocsType.JAVADOC:
+                sb.append("javadocs");
+                break;
+            case DocsType.SOURCES:
+                sb.append("sources");
+                break;
+            case DocsType.USER_MANUAL:
+                sb.append("a user manual");
+                break;
+            case DocsType.SAMPLES:
+                sb.append("samples");
+                break;
+            case DocsType.DOXYGEN:
+                sb.append("doxygen documentation");
+                break;
+            default:
+                sb.append("documentation of type '").append(name).append("'");
+        }
+    }
+
+    private static String toName(Object category) {
```

Comment:
:x: This is not only used for the category attribute. Parameter should be of type `Named`, the parameter should be named something like `attributeValue` and always return `attributeValue.getName()`.

Reply:
Same comment as above